// @polkadot/api-derive/staking/query.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.query = query;
exports.queryMulti = queryMulti;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _rxjs = require("rxjs");

var _operators = require("rxjs/operators");

var _util = require("@polkadot/util");

var _util2 = require("../util");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function parseSessionIds(stashId, queuedKeys, nextKeys) {
  const sessionIds = (queuedKeys.find(([currentId]) => currentId.eq(stashId)) || [undefined, []])[1];
  const nextSessionIds = nextKeys.unwrapOr([]);
  return {
    nextSessionIds,
    sessionIds
  };
}

function parseController(stashId, queuedKeys, [controllerIdOpt, nominatorsOpt, rewardDestination, validatorPrefs, nextKeys, exposure], stakingLedgerOpt) {
  const controllerId = controllerIdOpt.unwrapOr(null);
  const nominators = nominatorsOpt.unwrapOr(null);
  return controllerId && stakingLedgerOpt.isSome ? _objectSpread({
    accountId: stashId,
    controllerId,
    exposure,
    nominators: nominators ? Array.isArray(nominators) ? nominators[0].targets : nominators.targets : [],
    rewardDestination,
    stakingLedger: stakingLedgerOpt.unwrapOr(undefined),
    stashId,
    validatorPrefs: Array.isArray(validatorPrefs) ? validatorPrefs[0] : validatorPrefs
  }, parseSessionIds(stashId, queuedKeys, nextKeys)) : {
    accountId: stashId,
    nextSessionIds: [],
    sessionIds: []
  };
}

function retrievePrev(api, stashId) {
  return api.queryMulti([[api.query.staking.bonded, stashId], [api.query.staking.nominators, stashId], [api.query.staking.payee, stashId], [api.query.staking.validators, stashId], [api.query.session.nextKeys, [api.consts.session.dedupKeyPrefix, stashId]], [api.query.staking.stakers, stashId]]);
}

function retrieveCurr(api, stashIds, activeEra) {
  var _api$consts$session;

  return (0, _rxjs.combineLatest)([api.query.staking.bonded.multi(stashIds), api.query.staking.nominators.multi(stashIds), api.query.staking.payee.multi(stashIds), api.query.staking.validators.multi(stashIds), ((_api$consts$session = api.consts.session) === null || _api$consts$session === void 0 ? void 0 : _api$consts$session.dedupKeyPrefix) ? api.query.session.nextKeys.multi(stashIds.map(stashId => [api.consts.session.dedupKeyPrefix, stashId])) : api.query.session.nextKeys.multi(stashIds), api.query.staking.erasStakers.multi(stashIds.map(stashId => [activeEra, stashId]))]).pipe((0, _operators.map)(([controllerIdOpt, nominatorsOpt, rewardDestination, validatorPrefs, nextKeys, exposure]) => controllerIdOpt.map((controllerIdOpt, index) => [controllerIdOpt, nominatorsOpt[index], rewardDestination[index], validatorPrefs[index], nextKeys[index], exposure[index]])));
}

function retrieveControllers(api, optControllerIds) {
  const ids = optControllerIds.filter(opt => opt.isSome).map(opt => opt.unwrap());

  if (!ids.length) {
    return (0, _rxjs.of)(optControllerIds.map(() => api.registry.createType('Option<StakingLedger>')));
  }

  return api.query.staking.ledger.multi(ids).pipe((0, _operators.map)(optLedgers => {
    let offset = -1;
    return optControllerIds.map(opt => opt.isSome ? optLedgers[++offset] : api.registry.createType('Option<StakingLedger>'));
  }));
}
/**
 * @description From a stash, retrieve the controllerId and all relevant details
 */


function query(api) {
  return (0, _util2.memo)(accountId => api.derive.staking.queryMulti([accountId]).pipe((0, _operators.map)(([first]) => first)));
}

function queryMulti(api) {
  return (0, _util2.memo)(accountIds => accountIds.length ? (0, _rxjs.combineLatest)([api.query.session.queuedKeys(), api.derive.session.indexes()]).pipe((0, _operators.switchMap)(([queuedKeys, {
    activeEra
  }]) => {
    const stashIds = accountIds.map(accountId => api.registry.createType('AccountId', accountId));
    return ((0, _util.isFunction)(api.query.staking.erasStakers) ? retrieveCurr(api, stashIds, activeEra) : (0, _rxjs.combineLatest)(stashIds.map(stashId => retrievePrev(api, stashId)))).pipe((0, _operators.switchMap)(results => retrieveControllers(api, results.map(([optController]) => optController)).pipe((0, _operators.map)(stakingLedgerOpts => stashIds.map((stashId, index) => parseController(stashId, queuedKeys, results[index], stakingLedgerOpts[index]))))));
  })) : (0, _rxjs.of)([]));
}
